src/meta_schedule/utils.cc

```diff
@@ -0,0 +1,155 @@
+      if (i_block_idx == -1) {
+        i_block_idx = i;
+      }
+    }
+    if (IsThreadIdx(thread_scope)) {
+      if (i_thread_idx == -1) {
+        i_thread_idx = i;
+      }
+    }
+    if (loop->kind != tir::ForKind::kSerial) {
+      if (i_multi_child == -1) {
+        i_multi_child = i;
+      }
+    }
+    if (!IsSingleStmt(loop->body)) {
+      if (i_multi_child == -1) {
+        i_multi_child = i + 1;
+      }
+    }
+    if (tir::GetLoopIterType(loop_sref) == IterVarType::kDataPar) {
+      if (i_spatial_loop == i - 1) {
+        ++i_spatial_loop;
+      }
+    }
+  }
+  if (i_multi_child == -1) {
+    i_multi_child = n;
+  }
+  if ((i_block_idx != -1 && i_thread_idx != -1) || i_spatial_loop == -1) {
+    return BindType::kNoBind;
+  } else if (i_block_idx != -1 && i_thread_idx == -1) {
+    ICHECK(false) << "Unsupported case, where blockIdx is bound but threadIdx is not";
+    throw;
+  } else if (i_block_idx == -1 && i_thread_idx != -1) {
+    *fuse_first_num = std::min(std::min(i_multi_child, i_thread_idx), i_spatial_loop + 1);
+    return BindType::kBindBlock;
+  } else {  // i_block_idx == -1 && i_thread_idx == -1
+    *fuse_first_num = std::min(i_multi_child, i_spatial_loop + 1);
+    return BindType::kBindBlockThread;
+  }
+}
+
+Schedule BindThreadsForUnboundBlock(const Schedule& sch,      //
+                                    const BlockRV& block_rv,  //
+                                    int max_threadblock,      //
+                                    int max_num_threads,      //
+                                    Array<Integer> thread_extents) {
+  tir::StmtSRef block_sref = sch->GetSRef(block_rv);
+
+  int fuse_first_num = 0;
+  tir::BindType bind_type = tir::GetBindType(block_sref, &fuse_first_num);
+  if (bind_type == tir::BindType::kNoBind) {
+    return {sch};
+  }
+
+  Array<LoopRV> loop_rvs = sch->GetLoops(block_rv);
+  LoopRV fused = sch->Fuse({loop_rvs.begin(), loop_rvs.begin() + fuse_first_num});
+  if (bind_type == tir::BindType::kBindBlock) {
+    sch->Bind(fused, "blockIdx.x");
+  } else if (bind_type == tir::BindType::kBindBlockThread) {
+    int64_t extent_size = int64_t(1) << 60;
+    if (const int64_t* extent_ptr = tir::GetLoopIntExtent(sch->Get(fused).get())) {
+      extent_size = *extent_ptr;
+    }
+
+    Array<Integer> updated_extents;
+    for (const Integer extent : thread_extents) {
+      if (extent->value <= extent_size) updated_extents.push_back(extent);
+    }
+
+    if (extent_size <= max_threadblock * max_num_threads) {
+      tir::ExprRV factor;
+      if (updated_extents.empty()) {
+        factor = Integer(std::min(static_cast<int64_t>(max_num_threads), extent_size));
+      } else if (updated_extents.size() == 1) {
+        factor = updated_extents[0];
+      } else {
+        // Sample a factor
+        int n = updated_extents.size();
+        Array<FloatImm> probs(n, FloatImm(DataType::Float(64), 1.0 / n));
+        factor = sch->SampleCategorical(updated_extents, probs);
```

Comment:
so does it imply that we only sample when the loop extent <= max_threadblock * max_num_threads?

Reply:
That's it. Since if `loop extent >= max_threadblock * max_num_threads`, there are enough parallelism degrees, and the perf influence is limited.